### ISSUE TITLE
Doesn't TunnelKit support Network Extensions of type System Extension? (macOS)

### ISSUE THREAD

**Issue by zooxop** [passepartoutvpn/tunnelkit#319]
Hello. :)
I have a question.

I'm trying to distribute a macOS app I created with a .dmg (or .pkg) extension signed with a Developer ID Installer type certificate.
So, I found out that I can use a Network Extension of type System Extension to sign and distribute an app with a Developer ID type certificate, and I tried the following.

1. First, I developed an app that can tunnel to WireGuard server using Network Extension of Application Extension type. (Refer to the Demo project)
2. Add a new Network Extension of System Extension type and modify the settings and code according to the same procedure.
3. I tried to tunnel but failed and found the following error log in the Console.
> neagent(com.apple.networkextension) `NEAgentSession: failed to create the delegate`

So, I'm currently assuming that TunnelKit doesn't support the System Extension type.

Am I correct?

And if so, are there any plans to support the System Extension type?

**Comment by keeshux** [comment]
Hello, what documentation are you referring to for System Extension? I'm interested.

**Comment by zooxop** [comment]
I referenced a lot of pages, but here are the top ones I took note of.

1. Network Extension Entitlement Page
    1. [https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_networking_networkextension](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_networking_networkextension)
    2. additional : [https://developer.apple.com/documentation/systemextensions](https://developer.apple.com/documentation/systemextensions)
2. Developer Forums
    1. [https://developer.apple.com/forums/thread/125508](https://developer.apple.com/forums/thread/125508)

---

To summarize,

1. add a Network Extension of type System Extension.
2. Add `-systemextension` to the value of the Network Extension property in the `.entitlement` file.
3. replace Provisioning Profiles with Developer ID type.
4. Code sign & notarize.

I searched a lot of references and tried a lot of variations of the above method, as well as other minor variables, without success.

I asked because I wasn't sure if my method was wrong or if TunnelKit doesn't support the System Extension type.

Anyway, for now, I've decided to reorient my development towards deploying to the App Store wit h App Extension. 😂

**Comment by keeshux** [comment]
> I searched a lot of references and tried a lot of variations of the above method, as well as other minor variables, without success.

Please define "without success".

Thing is, originally, NetworkExtension was only meant for App Store deployments. Then I was told about this System Extension variation that would allow distribution of Developer ID builds, but never found the time to look into it. Also, there was _even_ less documentation than now at that time, hence my question out of curiosity.

Do not hesitate to share your findings here, I'm sure they will be of great help.

**Comment by zooxop** [comment]
> Please define "without success".

What I meant was that I couldn't run the `startTunnel()` method of the `PacketTunnelProvider`


> Do not hesitate to share your findings here, I'm sure they will be of great help.

Sure. I'm happy to share what I know if it can help in some small way.

However, I've been trying a lot of different things to figure out why I can't run `startTunnel()`, and I haven't kept detailed notes along the way.

So, if you keep this issue open for a while, I'll try to organize what I know and post it one by one.

Thank you. :)

**Comment by keeshux** [comment]
Sure, I'll leave this open long-term, unless I eventually reject the implementation for whatever reason. In that case, I'll let you know anyway.

Thank you very much!

**Comment by felstell** [comment]
I would like to join the discussion because I have encountered exactly the same problem as @zooxop. I don't have any specific information regarding this issue yet, but I'm willing to help if needed

**Comment by zooxop** [comment]
@felstell 

Great!

Please feel free to share any workarounds you have found for this issue, and I will do the same.

Unfortunately, since I opened this issue, I have decided to distribute the app I am working on through the App Store. As a result, I am currently dedicating all of my time to developing it, and I have not yet had a chance to explore other solutions. However, I did come across a project that could be a helpful reference for you:

https://github.com/eduvpn/mac-provisioning

At some point in the future, I plan to investigate how to use System Extensions further, using the above project as a guide. You might find it useful to check it out as well.

In any case, if you discover any additional insights, please share them here. It would be greatly appreciated.

Best of luck!
